[nanii] 2주차 midpoint

bread1022 opened the pull request:
# NANII - FE2. NEWS STAND

## 학습계획서

## 📌 학습 키워드
- [x] Flux

## 🔎 기능 요구 사항
- [ ] main 전체언론사 - 그리드 구독/해지
- [ ] main 전체언론사 - 리스트 UI
  - [ ] 화면구성
  - [ ] 리스트탭 UX
  - [ ] 프로그래스바 애니메이션
- [x] store부분을 view에서 분리
   - [ ] store와 view 섞이지 않도록 구현
   - [ ]  모듈간 의존성 제어

## 📝 주요 개발 구현 내용
[Notion-link](https://bebetter-nanii.notion.site/News-stand-f366288cc7304a9ca71dfefae5a7155e)

## 🤯 고민한 점, 느낀점
- 상태 전달 경로
- 역할에 대한 정의 내리기
- 데이터가 필요한 영역별로 Store 생성, 관리하고싶은데 상위 Store(전체데이터)에서 하위 Store에 어떻게 액션을 전달해야할지, 어떻게 데이터를 받을지, 어떤식으로 참조해야할지 정리가 안되서 우선 매개변수로  데이터 전달받아 처리했습니다. 리스트 영역을 구현하게 되면 State, action 처리를 다양하게 해야될 텐데 조금 더 고민해봐야겠습니다..

bread1022 committed:
feat: fetch util function 구현

- fetch 방법 아직 생각 중이지만 우선 유틸함수만 생성해놓음

bread1022 committed:
feat: rollingBar time diff 1sec 적용

- hover시 양쪽 동시 멈춤

bread1022 committed:
feat: main 태그 내부에 sortButtons 생성하는 함수 구현

bread1022 committed:
refactor: view 를 레이아웃으로 나눠서 관리

bread1022 committed:
feat: data Store class 생성 코드 구현

ahnlook commented on src/js/components/main/viewContainer/mainView.js:
```diff
@@ -0,0 +1,28 @@
+    CONSTANTS['ALL_PRESS'],
+    CONSTANTS['SUBSCRIBED_PRESS'],
+  );
+  const viewContainer = createEl('div', 'view-container');
```
class name까지 주는 유틸 만든 거 좋아보여용 :)

ahnlook commented on src/js/core/Store.js:
```diff
@@ -0,0 +1,31 @@
+    return this.#state;
+  }
+
+  subscribe(fn) {
```
`fn`은 어떤 줄임말인가용!

ahnlook commented on src/js/stores/dataStore.js:
```diff
@@ -0,0 +1,36 @@
+
+const dataReducer = (state, actionKey, { responseData }) => {
+  switch(actionKey) {
+    case 'DELIVER_DATA':
```
case가 하나인데 switch를 쓰신 이유가 있나용! 추가 될 예정인건가요?

ahnlook commented on src/js/stores/viewStore.js:
```diff
@@ -0,0 +1,21 @@
+const viewReducer = (state, actionKey, { data }) => {
+  switch(actionKey) {
+    case 'SET_GRIDVIEW':
+      return { ...data };
```
제가 store 개념을 잘 몰라서, store는 data를 전달하는 역할을 담당하고 있는건가요? 나니 store 파일을 보면 `return { some data }`만 있는 것 같아서 궁금했어요!

ahnlook commented on src/js/utils/fetch.js:
```diff
@@ -0,0 +1,8 @@
+const fetchData = async (...url) => {
```
fetch를 어디에 둬야할지 저도 고민 많이 했는데 아예 파일을 빼는 것도 좋아보여용! 이것도 패턴? 디자인의 일부인 건가요?

ahnlook commented on src/js/core/Store.js:
```diff
@@ -0,0 +1,31 @@
+  #listeners = [];
+  #reducer;
+
+  constructor(state, reducer) {
```
`reducer`에는 보통 어떤 것들이 들어오는 건가요?

ahnlook commented on src/js/stores/dataStore.js:
```diff
@@ -0,0 +1,36 @@
+  }
+}
+
+const dataReducer = (state, actionKey, { responseData }) => {
```
개인적으로 요즘 객체로 parameter 넘기는 거에 꽂혀서, 이렇게 받아서 처리하는 것도 깔끔할 것 같아용 :)
```
const dataReducer = (dataInformation) => {...}

dataInformation {
    state: // ... ,
    actionKey: // ... ,
    reponseData: // ...
}
```

ahnlook reviewed (commented):
코드가 깔끔해서 어려움 없이 이해했어요! 배워갈 부분이 많았습니당 :)

bread1022 committed:
feat: auto Rolling bar data 동적 구현 완료

- <li> Tag 를 2개만 넣고 textContent 를 바꿔치기 하는 로직으로 구현하고 싶었는데 만들어놓은 템플릿에 적용하기 어려워 reduce 메소드 사용해서 우선 구현함

bread1022 committed:
feat: Data Store 적용

- Data store 에서 모든 데이터를 패치한 뒤
- 롤링영역, 뷰영역 각각의 스토어에 생성하여 전달하고자 했으나 아직 설계중이므로 우선 매개변수로 데이터를 전달함

bread1022 committed:
refactor: Store 관련 코드 삭제

- reducer, dispatch, state 구현 완료 후 적용예정

crongro reviewed (changes requested):
상단에서 데이터를 가져오는 등의 비즈니스 작업을 해야 한다면 app 안에서 하는것이 맞긴하고요. 지금처럼 app 함수안에서 다 하지말고 데이터패칭하는 함수를 app.js안에 두고 해도 되긴하죠.

의도를 알수 없는 코드들은 이런거에요.
responseData[0] -> 0이 뭘까.. 'left', 'right' -> 왼쪽, 오른쪽..?뭐지?



datastore 의 data는 무엇인지 알기가 너무 어려워요. 의도를 자세히...

dataReducer 의 switch 문은 case가 하나라서..그냥 조건문으로 해도 되겠어요. 케이스가 더 늘어난다면 switch 가 좋을것 같고요.



components 아래 디렉토리 하나에 하나의 파일로 유지될 거 같으면 디렉토리가 없어도 되긴해요. 그 아래 하위 디렉토리가 있거나 여러파일이 있다면 디렉토리가 필요할 수도 있고요.



core, store, components 라는 역할로 구조를 만든 부분이 좋네요. 

리뷰참고해서 조금 수정해보세요.

bread1022 committed:
refactor: 변수로 의도를 명확하게 확인할 수 있도록 코드 수정

- data store의 역할이 fetch한 뒤 other store에 전달하는 것밖에 없어서 제거했습니다.
- 추후에 역할이 뚜렷해지면 생성할 예정입니다.

bread1022 committed:
feat: grid view Class 작성

- 버튼 클릭시, 그리드와 버튼 재렌더링하는 메소드 적용

bread1022 committed:
refactor: 변수명, 함수명 변경

crongro commented on src/js/components/main/viewContainer/grid/grid.js:
```diff
@@ -0,0 +1,47 @@
+
+  moveToPage() {
+    this.viewContainer.addEventListener('click', ({ target }) => {
+      if(!target.closest('a')) return;
```
a로 체크하는것도 좋은데 좀더 구체적인 정보를 가지고 판단해도 좋죠.

crongro commented on src/js/components/main/viewContainer/grid/grid.js:
```diff
@@ -0,0 +1,47 @@
+    const GRID_COUNT = 24;
+    const gridArea = createEl('div', 'grid-area');
+    const pressLogoData = this.#gridData.slice(GRID_COUNT * this.page, GRID_COUNT * (this.page + 1));
+    gridArea.innerHTML = pressLogoData.map((press) => `<div><img src=${press.mediaInfo.pressLogo}></div>`).join('');
```
map.join() 은 reduce 로  대체할수있죠.

crongro reviewed (approved):
수고하셨어요~ 머지해요.
